core/integration/container_test.go

```diff
@@ -0,0 +1,623 @@
+package core
+
+import (
+	"context"
+	"testing"
+
+	"github.com/Khan/genqlient/graphql"
+	"github.com/moby/buildkit/identity"
+	"github.com/stretchr/testify/require"
+	"go.dagger.io/dagger/engine"
+	"go.dagger.io/dagger/internal/testutil"
+)
+
+func TestContainerScratch(t *testing.T) {
+	t.Parallel()
+
+	res := struct {
+		Container struct {
+			ID     string
+			Rootfs struct {
+				Contents []string
+			}
+		}
+	}{}
+
+	err := testutil.Query(
+		`{
+			container {
+				id
+				rootfs {
+					contents
+				}
+			}
+		}`, &res, nil)
+	require.NoError(t, err)
+	require.Empty(t, res.Container.ID)
+	require.Empty(t, res.Container.Rootfs.Contents)
+}
+
+func TestContainerFrom(t *testing.T) {
+	t.Parallel()
+
+	res := struct {
+		Container struct {
+			From struct {
+				Rootfs struct {
+					File struct {
+						Contents string
+					}
+				}
+			}
+		}
+	}{}
+
+	err := testutil.Query(
+		`{
+			container {
+				from(address: "alpine:3.16.2") {
+					rootfs {
+						file(path: "/etc/alpine-release") {
+							contents
+						}
+					}
+				}
+			}
+		}`, &res, nil)
+	require.NoError(t, err)
+	require.Equal(t, res.Container.From.Rootfs.File.Contents, "3.16.2\n")
+}
+
+func TestContainerExecExitCode(t *testing.T) {
+	t.Parallel()
+
+	res := struct {
+		Container struct {
+			From struct {
+				Exec struct {
+					ExitCode int
+				}
+			}
+		}
+	}{}
+
+	err := testutil.Query(
+		`{
+			container {
+				from(address: "alpine:3.16.2") {
+					exec(args: ["true"]) {
+						exitCode
+					}
+				}
+			}
+		}`, &res, nil)
+	require.NoError(t, err)
+	require.Equal(t, res.Container.From.Exec.ExitCode, 0)
+
+	/*
+		It's not currently possible to get a nonzero exit code back because
+		Buildkit raises an error.
+
+		We could perhaps have the shim mask the exit status and always exit 0, but
+		we would have to be careful not to let that happen in a big chained LLB
+		since it would prevent short-circuiting.
+
+		We could only do it when the user requests the exitCode, but then we would
+		actually need to run the command _again_ since we'd need some way to tell
+		the shim what to do.
+
+		Hmm...
+
+		err = testutil.Query(
+			`{
+				container {
+					from(address: "alpine:3.16.2") {
+						exec(args: ["false"]) {
+							exitCode
+						}
+					}
+				}
+			}`, &res, nil)
+		require.NoError(t, err)
+		require.Equal(t, res.Container.From.Exec.ExitCode, 1)
```

Comment:
Opened https://github.com/dagger/dagger/issues/3192 to track this. It also affects the current `core { exec { exitCode } }`.